### PR TITLE
WFLY-9860 Infinispan store implementation based on remote-cache-container configuration

### PR DESCRIPTION
**Pull request by rhusar**
Jira
https://issues.jboss.org/browse/WFLY-9860

Requires
https://issues.jboss.org/browse/WFLY-6634
https://github.com/wildfly/wildfly/pull/9355

### CODE REVIEW

**Comment by rhusar** [comment]
Please review @pferraro

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreBuilder.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    public Builder<PersistenceConfiguration> configure(OperationContext context, ModelNode model) throws OperationFailedException {
+        String remoteCacheContainerName = REMOTE_CACHE_CONTAINER.resolveModelAttribute(context, model).asString();
+        this.remoteCacheContainer = new InjectedValueDependency<>(InfinispanRequirement.REMOTE_CONTAINER.getServiceName(context, remoteCacheContainerName), RemoteCacheContainer.class);
+        return super.configure(context, model);
+    }
+
+    @Override
+    public ServiceBuilder<PersistenceConfiguration> build(ServiceTarget target) {
+        ServiceBuilder<PersistenceConfiguration> builder = super.build(target);
+        this.remoteCacheContainer.register(builder);
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreBuilder.java]: return this.remoteCacheContainer.register(super.build(target));

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreConfigurationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public HotRodStoreConfigurationBuilder remoteCacheContainer(RemoteCacheContainer remoteCacheContainer) {
+        this.attributes.attribute(REMOTE_CACHE_CONTAINER).set(remoteCacheContainer);
+        return this;
+    }
+
+    @Override
+    public HotRodStoreConfiguration create() {
+        return new HotRodStoreConfiguration(this.attributes.protect(), this.async.create(), this.singletonStore.create());
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreConfigurationBuilder.java]: Don't bother passing singletonStore.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * @author Radoslav Husar
+ */
+@BuiltBy(HotRodStoreConfigurationBuilder.class)
+@ConfigurationFor(HotRodStore.class)
+public class HotRodStoreConfiguration extends AbstractStoreConfiguration {
+
+    static final AttributeDefinition<RemoteCacheContainer> REMOTE_CACHE_CONTAINER = AttributeDefinition.builder("remoteCacheContainer", null, RemoteCacheContainer.class).build();
+
+    public HotRodStoreConfiguration(AttributeSet attributes, AsyncStoreConfiguration async, @SuppressWarnings("deprecation") SingletonStoreConfiguration singletonStore) {
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreConfiguration.java]: Remove singletonStore from constructor.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreResourceDefinition.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ * /subsystem=infinispan/cache-container=X/cache=Y/store=hotrod
+ *
+ * @author Radoslav Husar
+ */
+public class HotRodStoreResourceDefinition extends StoreResourceDefinition {
+
+    public static final PathElement PATH = pathElement("hotrod");
+
+    public enum Attribute implements org.jboss.as.clustering.controller.Attribute {
+        REMOTE_CACHE_CONTAINER,
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreResourceDefinition.java]: Parameterize using name, type, and capability reference.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java [hunk 5]
```diff
@@ -78,18 +78,19 @@ public RemoteCacheContainerBuilder configure(OperationContext context, ModelNode
...
         Consumer<RemoteCacheManager> destroyer = remoteCacheManager -> {
             remoteCacheManager.stop();
-            InfinispanLogger.ROOT_LOGGER.debugf("%s remote cache container stopped", this.name);
+            InfinispanLogger.ROOT_LOGGER.remoteCacheContainerStopped(this.name);
         };
 
         Service<RemoteCacheContainer> service = new SuppliedValueService<>(mapper, supplier, destroyer);
         ServiceBuilder<RemoteCacheContainer> builder = target.addService(this.getServiceName(), service)
-                .setInitialMode(ServiceController.Mode.ON_DEMAND)
+                // Workaround org.infinispan.client.hotrod.RemoteCacheContainer not being restartable
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: We need to figure out why this is and come up with a workaround.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    @Override
+    public RemoteCacheContainerBuilder configure(OperationContext context, ModelNode model) throws OperationFailedException {
+        this.configuration = new InjectedValueDependency<>(InfinispanRequirement.REMOTE_CONTAINER_CONFIGURATION.getServiceName(context, this.name), Configuration.class);
+        return this;
+    }
+
+    @Override
+    public ServiceBuilder<RemoteCacheContainer> build(ServiceTarget target) {
+        Function<RemoteCacheManager, RemoteCacheContainer> mapper = remoteCacheManager -> new ManagedRemoteCacheContainer(this.name, remoteCacheManager);
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: You can remove this lambda by having RemoteCacheContainerBuilder implement Function<RemoteCacheManager, RemoteCacheContainer>

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+        Supplier<RemoteCacheManager> supplier = () -> {
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: You can remove this lambda by having RemoteCacheContainerBuilder implement Supplier\<RemoteCacheManager\>

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            Configuration configuration = this.configuration.getValue();
+            RemoteCacheManager remoteCacheManager = new RemoteCacheManager(configuration);
+            remoteCacheManager.start();
+            InfinispanLogger.ROOT_LOGGER.debugf("%s remote cache container started", this.name);
+            return remoteCacheManager;
+        };
+
+        Consumer<RemoteCacheManager> destroyer = remoteCacheManager -> {
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: You can remove this lambda by having RemoteCacheContainerBuilder implement Consumer\<RemoteCacheManager\>

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            remoteCacheManager.stop();
+            InfinispanLogger.ROOT_LOGGER.remoteCacheContainerStopped(this.name);
+        };
+
+        Service<RemoteCacheContainer> service = new SuppliedValueService<>(mapper, supplier, destroyer);
+        ServiceBuilder<RemoteCacheContainer> builder = target.addService(this.getServiceName(), service)
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: This needs to be wrapped using AsynchronousServiceBuilder(...) as the service start/stop methods are blocking.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStore.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+import org.wildfly.clustering.infinispan.spi.RemoteCacheContainer;
+
+/**
+ * Simple implementation of Infinispan {@link AdvancedLoadWriteStore} configured with a started container-managed {@link RemoteCacheContainer}
+ * instance. Does not perform wrapping entries in Infinispan internal objects, this stores "raw" values.
+ *
+ * @author Radoslav Husar
+ */
+@Store(shared = true)
+public class HotRodStore<K, V> implements AdvancedLoadWriteStore<K, V> {
```
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStore.java]: Either we need to append the name of the embedded cache to the cache key - or we need to create a new remote cache using the name of the embedded cache based on the default cache config.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreConfigurationBuilder.java (hunk 2 above)
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreConfigurationBuilder.java]: Sadly, still needed.

```
Caused by: java.lang.NullPointerException
        at org.infinispan.configuration.cache.SingletonStoreConfigurationBuilder.read(SingletonStoreConfigurationBuilder.java:97)
```

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreConfiguration.java (hunk 3 above)
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreConfiguration.java]: Sadly, still needed.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreResourceDefinition.java (hunk 4 above)
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStoreResourceDefinition.java]: Fixed.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java (hunk 5 above)
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: Fixed now with https://issues.jboss.org/browse/ISPN-8918 (thanks!)

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java (hunk 6 above)
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: Fixed in https://github.com/wildfly/wildfly/pull/9355

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java (hunk 7 above)
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: Fixed in https://github.com/wildfly/wildfly/pull/9355

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java (hunk 9 above)
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: Fixed in https://github.com/wildfly/wildfly/pull/9355

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStore.java (hunk 10 above)
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/HotRodStore.java]: Fixed with new Infinispan API org.infinispan.client.hotrod.RemoteCacheManagerAdmin#getOrCreateCache(java.lang.String, java.lang.String) defining a cache of the cache name using default configuration.

**Pull request merged**